src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```diff
@@ -0,0 +1,139 @@
+                new InvalidRequestException(
+                        InvalidRequestException.ErrorCode.UNSUPPORTED_REQUEST_METHOD);
+        log.info("Token endpoint does not support GET request." );
+        response.setStatus(exception.getHttpCode());
+        response.setContentType("application/json;charset=UTF-8");
+        response.getWriter().println(
+                OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+        response.getWriter().flush();
+    }
+
+    protected void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException {
+        try {
+            TokenEndpointRequestValidator.validatePost(request);
+        } catch(OAuth2Exception exception) {
+            log.info(
+                    "Failed in validating Post request in Token Endpoint." +
+                            "Error Type: " + exception.getErrorType() +
+                            "Description: " + exception.getErrorDescription());
+            response.setStatus(exception.getHttpCode());
+            response.setContentType("application/json;charset=UTF-8");
+            response.getWriter().println(
+                    OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+            response.getWriter().flush();
+            return;
+        }
+
+        OAuth2Request oauth2Request = parseOAuth2RequestFromHttpRequest(request);
+    }
+
+    /**
+     * Should be called after all validation in doPost function.
+     */
+    private OAuth2Request parseOAuth2RequestFromHttpRequest(HttpServletRequest request) {
+
+        String grantType = request.getParameter(OAuth2ParameterNames.GRANT_TYPE);
+
+        Preconditions.checkArgument(
+                OAuth2Utils.getClientSession(request).getClient().isPresent(),
+                "Client should have been set in client filter!");
+
+        OAuth2Request.Builder oauth2RequestBuilder = OAuth2Request.newBuilder();
+        oauth2RequestBuilder.getRequestAuthBuilder().setClientId(
+                            OAuth2Utils.getClientSession(request).getClient().get().getClientId());
+        oauth2RequestBuilder.getRequestBodyBuilder()
+                .setGrantType(grantType).setResponseType("token");
+
+        if (grantType.equals("authorization_code")) {
+            oauth2RequestBuilder.getRequestAuthBuilder().setCode(
+                    request.getParameter(OAuth2ParameterNames.CODE));
+            try {
+                oauth2RequestBuilder.getAuthorizationResponseBuilder().setRedirectUri(
+                        URLDecoder.decode(
+                                request.getParameter(OAuth2ParameterNames.REDIRECT_URI),
+                                "utf-8"));
+            } catch (UnsupportedEncodingException e) {
+                // This should never happen.
+                throw new IllegalStateException(
+                        "URL should be valid", e);
```

Comment:
I think this is InvalidRequestException. It can happen if the caller passes non-URL-encoded garbage as the redirect_uri query param. IllegalStateException makes more sense if you're guarding against a bug in your own codebase.

Reply:
I will validate it and throw InvalidRequestException in the validator.